scipy/optimize/cython_optimize/__init__.py

```diff
@@ -0,0 +1,630 @@
+Cython modules that can be imported using ``cimport``.
+
+.. py:module:: scipy.optimize.cython_optimize.zeros_struct
+
+``cython_optimize.zeros_struct``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The root-finders in this module use a callback that takes a double with the
+scalar independent variable as the 1st argument and a user defined ``struct``
+with any extra parameters as the 2nd. ::
+
+    double (*callback_type)(double, void*)
+
+This module can be imported using the following::
+
+    from scipy.optimize.cython_optimize cimport zeros_struct
+
+.. py:module:: scipy.optimize.cython_optimize.zeros_array
+
+``cython_optimize.zeros_array``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The root-finders in this module use a callback that takes an integer with the
+number of extra parameters as the 1st argument and an array of doubles with any
+extra parameters as the 2nd. Even if the integer is unused in your callback, it
+must still be in the signature, because an internal wrapper will use it to
+prepend the scalar independent variable to the array, which is then passed to
+your callback. Therefore, in your callback, the independent variable must be
+the first element in the array, followed by the extra parameters.  Also the
+maximum number of extra parameters is hard-coded as ``MAXARGS = 10``. ::
+
+    double (*callback_type_array)(int, double*)
+
+This module can be imported using the following::
+
+    from scipy.optimize.cython_optimize cimport zeros_array
+
+.. py:module:: scipy.optimize.cython_optimize.zeros_tuple
+
+``cython_optimize.zeros_tuple``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The root-finders in this module use a callback that takes a double with the
+scalar independent variable as the 1st argument and a Python tuple with any
+extra parameters as the 2nd. Therefore this signature is *not* safe to use with
+``nogil``. ::
+
+    double (*callback_type_tuple)(double, tuple)
+
+This module can be imported using the following::
+
+    from scipy.optimize.cython_optimize cimport zeros_tuple
+
+Available Functions
+-------------------
+There are seven available functions which are all implemented in pure C and
+exposed in modules matching the expected callback signature. These root-finding
+functions correspond to the functions available in ``scipy.optimize.zeros``.
+
+- :py:func:`~scipy.optimize.newton`
+- :py:func:`~scipy.optimize.bisect`
+- :py:func:`~scipy.optimize.ridder`
+- :py:func:`~scipy.optimize.brenth`
+- :py:func:`~scipy.optimize.brentq`
+
+Note, in :py:mod:`~scipy.optimize.cython_optimize`, Secant and Halley are
+separate from Newton. This differs from ``scipy.optimize.zeros`` in which they
+are all grouped into a single function, and differentiated by the presence of
+``fprime`` and ``fprime2``.
+
+:py:mod:`~scipy.optimize.cython_optimize.zeros_struct`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This module exposes root finders expecting extra arguments in a native C
+``struct``. See examples for usage.
```

Comment:
this doesn't look right. why not normal function signatures?

Reply:
Sorry, I don't understand what you mean. Are you referring to the Sphinx documentation markup? Does numpydoc work with Cython? Should I be adding docstrings to the `*.pyx` files?

I think I will just remove all of this documentation and try to emulate the documentation for [`scipy.special.cython_special`](https://docs.scipy.org/doc/scipy/reference/special.cython_special.html) and the [tutorial](https://docs.scipy.org/doc/scipy/reference/tutorial/special.html#cython-bindings-for-special-functions-scipy-special-cython-special) and hopefully that will be better.